File: Array8.js
// longest subarray with sum k (positives and negatives):
// given an array and a sum k , we need to print the length of the longest subarray that sums to k. : Brute force approach :native approach.

// function getLongestSubarray(a, k){
//     let n = a.length; 
//     let len = 0;

//     for(let i = 0; i < n; i++){
//            for(let j =i; j <n; j++){
//             let s = 0;
//             for(let K = i; K <= j; K++){
//                 s += a[K];
//             }
//             if(s === k)
//                  len = Math.max(len, j - i + 1);
//            }
//     }
//     return len;
// }

// let a = [-1, 1, 1];
// let k = 1;
// let len = getLongestSubarray(a, k);
// console.log("The length of the longest subarray is: ", len);




// better approach: using two loops: 
// function getLongestSubarray(a, k){
//     let n = a.length; 
//     let len =0;
//     for(let i = 0; i < n; i++){
//         let s = 0;
//          for(let j = i; j < n; j++){
//             s+= a[j];
//             if (s == k)
//                   len = Math.max(len, j - i + 1);
//          }
//     }
//     return len;
// }
// let a = [-1,1,1];
// let k = 1;
// let len = getLongestSubarray(a,k);
// console.log("The length of the longest subarray is : ", len);



// optimal approach: (using hashing ):
function getLongestSubarray(a, k){
    let n = a.length; 
    let preSumMap = new Map();
    let sum =0;
    let maxLen = 0;
    for(let i = 0; i< n; i++){
        sum += a[i];
        if(sum === k){
             maxLen = Math.max(maxLen, i+1);
        }

        let rem = sum - k ;
         
        if(preSumMap.has(rem)){
            let len = i - preSumMap.get(rem);
            maxLen = Math.max(maxLen, len);
        }
        if(!preSumMap.has(sum)){
            preSumMap.set(sum, i);
        }
    }
    return maxLen;
}

let a = [-1, 1, 1];
let k = 1;
let len = getLongestSubarray(a, k);
console.log("The length of the longest subarray is: ", len); 

File: Array8.cpp
// longest subarray with sum k (positives and negatives):
// given an array and a sum k , we need to print the length of the longest subarray that sums to k. : Brute force approach :native approach.
// #include<bits/stdc++.h>
// using namespace std;

// int getLongestSubarray(vector<int>& a, int k){
//     int n = a.size();

//     int len = 0; 
//     for(int i = 0; i< n; i++){
//          for(int j=i; j < n; j++){
//             int s = 0;
//             for(int K = i; K <= j; K++){
//                 s += a[K];
//             }
//             if(s == k)
//               len = max(len, j-i +1);
//          }
//     }
//     return len; 
// }

// int main(){
//     vector<int> a = {-1, 1, 1};
//     int k = 1;
//     int len = getLongestSubarray(a, k);
//     cout << "the length of the longest subarray is: " << len << "\n";
//     return 0;
// }



// better approach : 
// #include<bits/stdc++.h>
// using namespace std;

// int getLongestSubarray(vector<int>& a, int k ){
//     int n = a.size();
//     int len = 0;
//     for(int i = 0; i < n; i++){
//          int s = 0; 
//          for(int j = i; j < n; j++){
//             s += a[j];

//             if(s == k)
//               len = max(len, j - i + 1);
//          }
//     }
//     return len; 
// }

// int main(){
//      vector<int> a = {-1, 1, 1};
//      int k = 1;
//      int len = getLongestSubarray(a, k);
//      cout << "the length of the longest subarray is: " <<len << "\n";
//      return 0;
// }





//optimal approach: (using hashing): 
 #include<bits/stdc++.h>
 using namespace std;

 int getLongestSubarray(vector<int>& a, int k ){
    int n = a.size();
    map<int, int> preSumMap;
    int sum = 0;
    int maxLen = 0;
    for(int i = 0; i < n; i++){
         sum += a[i];

         if(sum == k){
             maxLen = max(maxLen, i + 1);
         }

         int rem = sum - k; 

         if(preSumMap.find(rem) != preSumMap.end()){
             int len = i - preSumMap[rem];
             maxLen = max(maxLen, len);
         }

         if(preSumMap.find(sum) == preSumMap.end()){
            preSumMap[sum] = i; 
         }
    }
    return maxLen; 
 }

 int main(){
   vector<int> a = {-1 , 1, 1};
   int k = 1;
   int len = getLongestSubarray(a, k);
   cout << "The length of the longest subarray is : " << len << "\n";
   return 0; 
 }